aspects/schema.go

```diff
@@ -0,0 +1,328 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+type parser interface {
+	Schema
+
+	// parseConstraints parses constraints for a type defined as a JSON object.
+	// Shouldn't be used with non-object/map type definitions.
+	parseConstraints(json.RawMessage) error
+}
+
+func ParseSchema(raw []byte) (*StorageSchema, error) {
+	var schemaDef map[string]json.RawMessage
+	err := json.Unmarshal(raw, &schemaDef)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse top level schema: must be a map")
+	}
+
+	schema := &StorageSchema{}
+	// TODO: check "types" here and parse the user-defined types
+
+	if _, ok := schemaDef["schema"]; !ok {
+		return nil, fmt.Errorf(`cannot parse top level schema: must have a "schema" constraint`)
+	}
+
+	schema.topLevel, err = schema.Parse(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	return schema, nil
+}
+
+type StorageSchema struct {
+	// topLevel is the schema for the top level map.
+	topLevel Schema
+}
+
+// Validate validates the provided JSON object
+func (s *StorageSchema) Validate(raw []byte) error {
+	return s.topLevel.Validate(raw)
+}
+
+func (s *StorageSchema) Parse(raw json.RawMessage) (parser, error) {
+	var typ string
+	var schemaDef map[string]json.RawMessage
+	var hasConstraints bool
+	if err := json.Unmarshal(raw, &schemaDef); err != nil {
+		var typeErr *json.UnmarshalTypeError
+		if !errors.As(err, &typeErr) {
+			return nil, err
```

Comment:
I think this covers all the relevant code paths (namely UnmarshalTypErrors since some schemas can have different definitions). I think adding tests for SyntaxErrors and things like that seems like it wouldn't add much value. But if there's something in particular that you think should be tested for, I can add it, just let me know